Add float64 kernels for Floor, Ceil, IsNaN

jbachurski opened the pull request:
### Description
This PR adds support for `float64` kernels in the latest versions of operators: Floor, Ceil and IsNaN.

### Motivation and Context
The lack of these kernels is non-trivial to work around and easily lead to performance losses when it is attempted. When equivalence with an existing implementation is required, precision is easily lost when casting to `float32` instead.

IsNaN is common when cleaning up data in an ML pipeline. Floor and Ceil have uses for discretising values and single-precision floats are insufficient to round well when values get larger than a few million.

According to my measurement this only increases the binary size by a few kilobytes (on the Python wheel of RelWithDebInfo).

Closes #13673 (Round already has float64 support)
Partially solves #8791 (Looks like there's parallel issues/PR open for Split, but it is also hard to work around and hence useful)

jbachurski commented:
@microsoft-github-policy-service agree

jbachurski commented:
> @jbachurski please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
>  > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="QuantCo"

tianleiwu commented on onnxruntime/core/providers/cpu/math/element_wise_ops.cc:
```diff
@@ -240,9 +240,11 @@ REG_ELEMENTWISE_TYPED_KERNEL(Neg, 13, int64_t, Neg);
 
 REG_ELEMENTWISE_VERSIONED_TYPED_KERNEL(Floor, 6, 12, float, Floor);
```
According to https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Floor-6, the double is supported since opset verison 6 (opset 13 Floor added bfloat16). I think the following line can be added:
``
REG_ELEMENTWISE_VERSIONED_TYPED_KERNEL(Floor, 6, 12, double, Floor);
``

JakubBachurskiQC replied:
Okay, I added the older-version kernels too!

tianleiwu commented:
This only adds cpu kernel. Shall we also add it to Cuda execution provider?

JakubBachurskiQC commented:
@tianleiwu Thanks for taking a look at the PR!

I took a short read through the CUDA execution provider and looks like both Floor and Ceil have `double` support on both versions of the operator. However, I don't think IsNaN is implemented at all there and I don't think I'll be able to implement or test this behaviour as I don't have a CUDA-capable device.

I added some basic tests as you asked for the new kernels. I also updated the kernels doc (I had some trouble building it though).

jbachurski commented:
> @jbachurski,
> 
> This only adds cpu kernel. Shall we also add it to Cuda execution provider?
> 
> May need build with --gen_doc to update documents.
> 
> Could you add a few test cases for this change?

@tianleiwu Thanks for taking a look at the PR!

I took a short read through the CUDA execution provider and looks like both Floor and Ceil have double support on both versions of the operator. However, I don't think IsNaN is implemented at all there and I don't think I'll be able to implement or test this behaviour as I don't have a CUDA-capable device.

I added some basic tests as you asked for the new kernels. I also updated the kernels doc (I had some trouble building it though).

tianleiwu commented:
/azp run Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,MacOS CI Pipeline

tianleiwu commented:
/azp run orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

tianleiwu reviewed (commented):
LGTM. Started CI pipelines...

jbachurski commented:
@tianleiwu Looks like the only failing job timed out at 3 hours. Any ideas way/how I could fix this? Or just rerun the check?

tianleiwu commented:
@jbachurski, I re-run the failed job.
